### PR TITLE
[BEAM-9228] Support further partition for FnApi ListBuffer

### PR DESCRIPTION
**Pull request by Hannah-Jiang**
A ListBuffer is partitioned when len(input) >= number of workers. This prevents parallelism for some transforms, like SDF. The PR supports further partitioning of the ListBuffer when len(input) < number of workers. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by Hannah-Jiang** [comment]
retest this please

**Comment by yifanzou** [comment]
retest this please

**Comment by Hannah-Jiang** [comment]
R: @robertwb 
Current PR passes all tests, except SDF ones. Can you please take a look to see I am on the correct track before I spend more time on investigation? Thank you.

**Comment by Hannah-Jiang** [comment]
errors: https://builds.apache.org/job/beam_PreCommit_Python_Commit/11130/testReport/

File: sdks/python/apache_beam/runners/portability/fn_api_runner.py [hunk 1]
```diff
@@ -937,6 +997,12 @@ def get_input_coder_impl(transform_id):
               process_bundle_descriptor.transforms[transform_id].spec.payload).
                       coder_id]].get_impl()
 
+    def get_input_coder(transform_id):
+      coder_id = beam_fn_api_pb2.RemoteGrpcPort.FromString(
+          process_bundle_descriptor.transforms[transform_id].spec.payload
+      ).coder_id
+      return context.coders[coder_id]
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/fn_api_runner.py]: This might fix your SDF issue as well. But if you use get_input_coder_impl everywhere, that might also fix things too.

File: sdks/python/apache_beam/runners/portability/fn_api_runner.py [hunk 2]
```diff
@@ -907,6 +961,12 @@ def get_buffer(buffer_id):
       if kind in ('materialize', 'timers'):
         # If `buffer_id` is not a key in `pcoll_buffers`, it will be added by
         # the `defaultdict`.
+        if buffer_id not in pcoll_buffers:
+          coder_id = beam_fn_api_pb2.RemoteGrpcPort.FromString(
+              process_bundle_descriptor.transforms[transform_id].spec.payload
+          ).coder_id
+          coder = context.coders[coder_id]
```
[the remark this answers is not in the document]
Comment (Hannah-Jiang) [reply sdks/python/apache_beam/runners/portability/fn_api_runner.py]: I have checked code for safe_coders, but still don't have clear ideas about it. What is the purpose of it? How do we handle for some coder_ids not part of safe_coders? Previously, I tried using safe_coders if the coder_id is in safe_coder, else return `coders[coder_id]`. This also pass tests, is it an option here?

File: sdks/python/apache_beam/runners/portability/fn_api_runner.py [hunk 3]
```diff
@@ -932,10 +998,14 @@ def get_buffer(buffer_id):
...
 
     def get_input_coder_impl(transform_id):
-      return context.coders[
-          safe_coders[beam_fn_api_pb2.RemoteGrpcPort.FromString(
-              process_bundle_descriptor.transforms[transform_id].spec.payload).
-                      coder_id]].get_impl()
+      coder_id = beam_fn_api_pb2.RemoteGrpcPort.FromString(
+          process_bundle_descriptor.transforms[transform_id].spec.payload).\
+        coder_id
+      assert coder_id is not None and coder_id != ''
```
Comment (Hannah-Jiang) [inline sdks/python/apache_beam/runners/portability/fn_api_runner.py]: `coder_id` can be retrieved from transform payload most cases, except one following case. It returns `''` coder_id. 
This cause `test_checkpoint_sdf` test fails. How do we need to handle this case? Other SDF tests should be fixed now.

```
spec {
  urn: "beam:transform:sdf_process_sized_element_and_restrictions:v1"
  payload: "\n\215\014\n beam:dofn:pickled_python_info:v1\032\350\013eNp9ldt31FQUxnNmBihpKSCorRdE8TLVOlBaLkXFy0AFI0NNEaJSYiY50wQyyXzJCVKcEVAzxfv9vf0XXcu1fPDBfc5MVy9QXmbNuezz/fa39zm5Wyy7XhCGFVv+6rabcEdwWyy2uA5t41oYO15/hVklTdPkfxRyFE2UrCGaOBtlTZ7QAXNeA9tcp+W4Prfr3GlWROJEaSNOmmnFjROun4lnIh3bc+zoYKBs6RRt283Yy0Ju29hpDfQmnCCioW7toGEriV2ephh8GHEji1wRxHTmUNlnKrwae/ySRNyVY9jE7rJRNDRjm1EyJqszC1qn2C4sayvaddYp3Rqn/4V2aYUlp9rF5WKbtUuNwgpj2n1tubSiMa1d8tgJrVPCnitau3iZZjxtTsOemsG62Fu29pGgmyUJj4Sd8FQkgcLBY9Y2WkmFkwjss3ZK15JF2w2doIn91nYaO60Wjzw8rhaDiPJp0iF4QuDJsrI55WEDI8oCHvYWR5XeOh2b/HVv8ARPqX0itnkzEHjaKva48IzAs0ou9RqV1iIO+IMXNFYd1kolprMhNswG2FABz1m712RsN84iQYcezPH8mMALJg65tl3PglDIuuirBdL9wUcW2zbXSGXhZ53Eaep4ceweXurg5bK1f1M2VOqbgUfKr/ha2S9Ze9c31+zqYvmRous0VyN0jOV4tYPXyv5OX1ceBlEgAifcULNxn3qI2qVA7VKgdpmrDgltmXmszVYKVPTXa4bWReUB6n4NUhy2BmnpYqORcmE60QLHkS4m/BF/tIuj/gF/3Pi7WtAYJsfI1SkTx7ZwdbxW6+B4TeCEiZPWsCxl/372qz29gbRIpLokJU5GnKdqOd7wD1sjazBrtJf6R7zZxVv+iAJe78HpLt4m0mnjH0X6jiR918R7W5BOS9IqkZ4xcbbX8q2QGnCmDyhvXtFgBLi9XViQcO8T3Lkx6Yl/egnnSWvG+Je0CvhAahkmPtxCa0ZqXSCtmomL1p5NRUiD2xyzG2QLJDvQu+wk/BEJm72LJbfO5bi0SvExUcwa/6mML0uKKyasLShmJcUnRPGpic/UpbNtL6bNuFqzdqlhGsYicpo8pcn5eVzLBGwTn1PXO+qBaslrYAce6tYhGq5vWLlSfqDjx+Bm9Rzewx5Al9MAvJ84o8QZJV6tDgitrY3KxBuU+II1KtFagXsj5J696aKn8HMEyhvJjes5bqx/NMJedBb14jeHo3nhGlPmRdK82ERrC/Oa0jyQeYmJ1DpIh04frfOTE8ecqfqkNzVVn6jzqSMnjx6ZPO5OT7j82AREjszEzRxfmLiVY7GD2xT/pYm2f7Xmk70d1cPqA2X7QSRSfPUwn2RinshaIX3G7qiQ8xflV+KcCrmrZoKolQl1Uop76ssWZ2Jt6mvlSZwEC0GEb5bwrcpgw1tE+xRDxeP0GDkiJm/zJXRNLNVq87i/hO9MfE8N8YOJH6khfurg5wf4f/HV5l+X8JuJ39P6PP7o4M8apf1X5X8g4Xjr0\001:\037ref_Coder_FastPrimitivesCoder_4"
}
inputs {
  key: "0"
  value: "ref_PCollection_PCollection_3_split"
}
outputs {
  key: "None"
  value: "ref_PCollection_PCollection_4"
}
unique_name: "SDF/Process"
environment_id: "ref_Environment_default_environment_1"
```

File: sdks/python/apache_beam/runners/portability/fn_api_runner.py [hunk 4]
```diff
@@ -287,43 +287,48 @@ def partition(self, n):
...
     else:
       if not self._grouped_output:
-        self._grouped_output = [[] for _ in range(n)]
-        coder_impl = self._input_coder.get_impl()
-        decoded_input = []
-        output_stream_list = []
-        for _ in range(n):
-          output_stream_list.append(create_OutputStream())
+        output_stream_list = [create_OutputStream() for _ in range(n)]
+        self._grouped_output = [output_stream.get() for output_stream
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner.py]: I meant you could put this instead of the loop at line 324.

File: sdks/python/apache_beam/runners/portability/fn_api_runner.py [hunk 5]
```diff
@@ -769,7 +774,7 @@ def _collect_written_timers_and_add_to_deferred_inputs(
         for windowed_key_timer in timers_by_key_and_window.values():
           windowed_timer_coder_impl.encode_to_stream(
               windowed_key_timer, out, True)
-        deferred_inputs[transform_id] = _ListBuffer(input_coder=coder)
+        deferred_inputs[transform_id] = _ListBuffer(coder_impl=coder.get_impl())
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner.py]: You can now revert the changes up at the top of the loop.

File: sdks/python/apache_beam/runners/portability/fn_api_runner.py [hunk 6]
```diff
@@ -932,10 +998,14 @@ def get_buffer(buffer_id):
       return pcoll_buffers[buffer_id]
 
     def get_input_coder_impl(transform_id):
-      return context.coders[
-          safe_coders[beam_fn_api_pb2.RemoteGrpcPort.FromString(
-              process_bundle_descriptor.transforms[transform_id].spec.payload).
-                      coder_id]].get_impl()
+      coder_id = beam_fn_api_pb2.RemoteGrpcPort.FromString(
+          process_bundle_descriptor.transforms[transform_id].spec.payload).\
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner.py]: Lint: don't use backslashes for continuation. (These days you can just run yapf to format your code, see https://cwiki.apache.org/confluence/display/BEAM/Python+Tips.)

File: sdks/python/apache_beam/runners/portability/fn_api_runner.py (hunk 3 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/fn_api_runner.py]: OK, in the SDF case we can get the coder of the preceding transform (the one that produces its input `ref_PCollection_PCollection_3_split`) which should always be a RemoteGrpcPort.

File: sdks/python/apache_beam/runners/portability/fn_api_runner.py [hunk 7]
```diff
@@ -994,7 +1069,13 @@ def input_for(transform_id, input_id):
...
         for other_input in data_input:
           if other_input not in deferred_inputs:
-            deferred_inputs[other_input] = _ListBuffer([])
+            outputs = process_bundle_descriptor.transforms[
+              other_input].outputs.values()
+            coder_id = process_bundle_descriptor.pcollections[
+              only_element(outputs)].coder_id
+            coder = context.coders[coder_id]
+            deferred_inputs[other_input] = _ListBuffer(
+                coder_impl=coder.get_impl())
```
Comment (Hannah-Jiang) [inline sdks/python/apache_beam/runners/portability/fn_api_runner.py]: As commented at L1082 (of the PR branch), deferred inputs cannot be parallel processed for now. Is it better to set coder_impl to None to reduce unnecessary processes for now and add it back later when parallel processing is supported for deferred_inputs?

**Comment by Hannah-Jiang** [comment]
retest it please

**Comment by Hannah-Jiang** [comment]
@robertwb, please take a look.

**Comment by Hannah-Jiang** [comment]
kindly ping. @robertwb

File: sdks/python/apache_beam/runners/portability/fn_api_runner.py [hunk 8]
```diff
@@ -932,10 +994,14 @@ def get_buffer(buffer_id):
...
     def get_input_coder_impl(transform_id):
-      return context.coders[
-          safe_coders[beam_fn_api_pb2.RemoteGrpcPort.FromString(
-              process_bundle_descriptor.transforms[transform_id].spec.payload).
-                      coder_id]].get_impl()
+      coder_id = beam_fn_api_pb2.RemoteGrpcPort.FromString(
+          process_bundle_descriptor.transforms[transform_id].spec.payload
+      ).coder_id
+      assert coder_id
+      if coder_id in safe_coders:
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner.py]: An alternative would be `safe_coders.get(coder_id, coder_id)` which means look up the key (first argument) and if it doesn't exist return the second argument. This would eliminate some of the repetition of logic between the lines as well.

File: sdks/python/apache_beam/runners/portability/fn_api_runner_test.py [hunk 9]
```diff
@@ -1603,23 +1603,6 @@ def restriction_size(self, element, restriction):
     return restriction.size()
 
 
-class FnApiRunnerSplitTestWithMultiWorkers(FnApiRunnerSplitTest):
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner_test.py]: Is there any loss of coverage in removing these tests? I can see how perhaps they're too strict.
